internal/agentdeployer/factory.go

```diff
@@ -80,23 +67,51 @@ func Factory(options FactoryOptions) (AgentDeployer, error) {
 		// FIXME: this docker-compose scenario contains both agent and service
 		return nil, nil
 	case "k8s":
-		if _, err := os.Stat(agentDeployerPath); err == nil {
-			opts := KubernetesAgentDeployerOptions{
-				Profile:        options.Profile,
-				DefinitionsDir: agentDeployerPath,
-				StackVersion:   options.StackVersion,
-				PolicyName:     options.PolicyName,
-				DataStream:     options.DataStream,
-				RunSetup:       options.RunSetup,
-				RunTestsOnly:   options.RunTestsOnly,
-				RunTearDown:    options.RunTearDown,
-			}
-			return NewKubernetesAgentDeployer(opts)
+		opts := KubernetesAgentDeployerOptions{
+			Profile:        options.Profile,
+			DefinitionsDir: agentDeployerPath,
+			StackVersion:   options.StackVersion,
+			PolicyName:     options.PolicyName,
+			DataStream:     options.DataStream,
+			RunSetup:       options.RunSetup,
+			RunTestsOnly:   options.RunTestsOnly,
+			RunTearDown:    options.RunTearDown,
 		}
+		return NewKubernetesAgentDeployer(opts)
 	}
 	return nil, fmt.Errorf("unsupported agent deployer (name: %s)", agentDeployerName)
 }
 
+func selectAgentDeployerType(options FactoryOptions) (string, string, error) {
+	devDeployPath, err := FindDevDeployPath(options)
+	if errors.Is(err, os.ErrNotExist) {
+		return "default", "", nil
+	}
+	if err != nil {
+		return "", "", fmt.Errorf("can't find \"%s\" directory: %w", options.DevDeployDir, err)
+	}
+
+	agentDeployerName, err := findAgentDeployer(devDeployPath)
+	if err != nil {
+		logger.Debugf("Not found any agent deployer, using default one")
+		return "default", "", nil
+	}
+	// if package defines `_dev/deploy/docker` folder to start their services, it should be
+	// using the default agent deployer`
+	if agentDeployerName == "docker" || agentDeployerName == "tf" {
+		return "default", "", nil
+	}
+
+	agentDeployerPath := filepath.Join(devDeployPath, agentDeployerName)
+	if info, err := os.Stat(agentDeployerPath); err != nil {
```

Comment:
Question: should this path be already ensure that exists ? 

`devDeployPath` should be a valid path returned by `FindDevDeployPath` but looking at the code it could be a file instead of a directory.

What about moving this check if that path is a folder into `FindDevDeployPath` method? 


The string returned by `findAgentDeployer` should be already a directory (maybe there is an edge case that I missed):

https://github.com/elastic/elastic-package/blob/f8153ab029d82c2107f3c6d0f6de5a9a478550d9/internal/agentdeployer/factory.go#L127